

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Documentation &mdash; Expression Store 0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="scripts" href="modules.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Expression Store
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">scripts</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Documentation</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Expression Store</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="modules.html">scripts</a> &raquo;</li>
        
      <li>Documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/database_preprocessor.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-database_preprocessor">
<span id="documentation"></span><h1>Documentation<a class="headerlink" href="#module-database_preprocessor" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="database_preprocessor.DataProcessor">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">database_preprocessor.</span></code><code class="sig-name descname"><span class="pre">DataProcessor</span></code><a class="headerlink" href="#database_preprocessor.DataProcessor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="database_preprocessor.DataProcessor.concat_gdc_meta">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">concat_gdc_meta</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inpath_gdc_meta</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.concat_gdc_meta" title="Permalink to this definition">¶</a></dt>
<dd><p>This method concatenates metadata, downloaded from GDC
:param inpath_gdc_meta: Path to directory, where GDC metadata has been saved
:return: DataFrame containing information about GDC metadata</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.concat_icgc_meta">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">concat_icgc_meta</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_dcc_manifest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.concat_icgc_meta" title="Permalink to this definition">¶</a></dt>
<dd><p>This method joins metadata from ICGC DCC (Donor, Sample, Project, Repository)
based on their matching columns
:param metadata_dest: Path to directory, where metadata has been downloaded to
:param path_to_dcc_manifest: Path to ICGC DCC manifest file
:return: returns DataFrames merged repository and manifest, sample, donor</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.concat_sra_meta">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">concat_sra_meta</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inpath_sra_meta_csv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">used_accessions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.concat_sra_meta" title="Permalink to this definition">¶</a></dt>
<dd><p>This method concatenates all downloaded metadata files from SRA, if they are in scope of the SRA files,
which have been processed with nf-core/rnaseq
:param inpath_sra_meta_csv: Path to directory, where the SRA metadata has been saved
:param used_accessions: SRA accessions, which have been used for analysis with rnaseq
:return: DataFrame containing all processed metadata from SRA</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.create_countinfo_raw_count_table">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">create_countinfo_raw_count_table</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rnaseq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.create_countinfo_raw_count_table" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates the countinfo table
:param df: DataFrame containing some f the countinfo information
:param rnaseq: Path to reuslts of nf-core/rnaseq
:param db_dest: Path do directory where files for database should be stored
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.create_expression_table">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">create_expression_table</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">expression_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw_count_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.create_expression_table" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates the expression csv fil
:param expression_df: expression DataFrame
:param raw_count_df: raw_count DataFrame from featureCounts
:param db_dest: Path to directory, where cdv files should be saved
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.create_gene_table">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">create_gene_table</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ensembl_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hgnc_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.create_gene_table" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates the gene_table csv files
:param ensembl_df: Ensembl DataFrame
:param hgnc_df: HUGO DataFrame
:param expression_df: Expression DataFrame
:param db_dest: Path to directory, where cdv files should be saved
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.create_pipeline_table">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">create_pipeline_table</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rnaseq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.create_pipeline_table" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates the pipeline table for the database
:param rnaseq: Path to reuslts of nf-core/rnaseq
:param db_dest: Path do directory where files for database should be stored
:return: None</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.create_project_layer_tables">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">create_project_layer_tables</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sra_meta_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">icgc_meta_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gdc_meta_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.create_project_layer_tables" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates tables for ProjectLayer
:param sra_meta_df: Metadata DataFrame from SRA
:param icgc_meta_df: Metadata DataFrame from ICGC DCC
:param gdc_meta_df: Metadata DataFrame from GDC
:param db_dest: Path to directory, where cdv files should be saved
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.process_ensembl_metadata">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">process_ensembl_metadata</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.process_ensembl_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>This method processes ensembl metadata from the processed gtf file
:param metadata_dest: Path to directory, where metadata has been downloaded to
:return: DataFrame containing information about Ensembl</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.process_hgnc_metadata">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">process_hgnc_metadata</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.process_hgnc_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>This method processes the metadata from HGNC
:param metadata_dest: Path to directory, where metadata has been downloaded to
:return: DataFrame containing information about HGNC</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.DataProcessor.stringtie_results_out">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">stringtie_results_out</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">res_inpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.DataProcessor.stringtie_results_out" title="Permalink to this definition">¶</a></dt>
<dd><p>This method extracts information from nf-core/rnaseq’s output
:param res_inpath: Path to rna-seq results folder
:return: returns list of DataFrames containing information from StringTie</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="database_preprocessor.Downloader">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">database_preprocessor.</span></code><code class="sig-name descname"><span class="pre">Downloader</span></code><a class="headerlink" href="#database_preprocessor.Downloader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="database_preprocessor.Downloader.download_ensembl_metadata">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">download_ensembl_metadata</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Downloader.download_ensembl_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>This method downloads data from Ensembl!
:param metadata_dest: Path to directory, where metadata has been downloaded to
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Downloader.download_gdc_file_metadata">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">download_gdc_file_metadata</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gdc_file_name_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Downloader.download_gdc_file_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>This method downloads all required Metadata from GDC
:param gdc_file_name_list: Unique list of GDC file names
:param metadata_dest: Path to directory, where metadata should be downloaded to
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Downloader.download_hgnc">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">download_hgnc</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Downloader.download_hgnc" title="Permalink to this definition">¶</a></dt>
<dd><p>This method downloads metadata from HGNC
:param metadata_dest: Path to directory, where metadata should be saved
:return: DataFrame containing information from HGNC</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Downloader.download_icgc_project_metadata">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">download_icgc_project_metadata</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">icgc_project_lst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Downloader.download_icgc_project_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>This method downloads all necessary metadata from ICGC DCC
:param icgc_project_lst: Unique List of Project Codes from ICGC DCC
:param metadata_dest: Path to directory, where metadata should be downloaded tod
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Downloader.download_sra_metadata">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">download_sra_metadata</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sra_acc_lst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Downloader.download_sra_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>This method downloads runInfo.csv from SRA depending on a list of SRA accessions
:param sra_acc_lst: List containing unique SRA Run accession
:param metadata_dest: Path to directory, where metadata should be downloaded to
:return:</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="database_preprocessor.Helper">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">database_preprocessor.</span></code><code class="sig-name descname"><span class="pre">Helper</span></code><a class="headerlink" href="#database_preprocessor.Helper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="database_preprocessor.Helper.create_csv">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">create_csv</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.create_csv" title="Permalink to this definition">¶</a></dt>
<dd><p>This method exports a DataFrame to a csv-File with no Index
:param df: DataFrame to be exported
:param file_name: File name for the exported csv-file
:param db_dest: Destination Directory, where csv should be saved
:return: None</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.create_csv_from_dask">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">create_csv_from_dask</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ddf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.create_csv_from_dask" title="Permalink to this definition">¶</a></dt>
<dd><p>This method exports a DASK DataFrame to a csv-File with no Index
:param ddf: dask DataFrame
:param file_name: File name for the exported csv-file
:param db_dest: Destination Directory, where csv should be saved
:return:</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.extract_icgc_run_id">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">extract_icgc_run_id</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.extract_icgc_run_id" title="Permalink to this definition">¶</a></dt>
<dd><p>This method splits a specific string, created by ICGC.The split truncates the Read suffix.
:param string: Any given string, which might needs be truncated.
:return: splitted String</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.extract_sample_id">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">extract_sample_id</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">string</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.extract_sample_id" title="Permalink to this definition">¶</a></dt>
<dd><p>This method splits sample_id strings, and truncates the substrings generated by StringTie
:param string: Any given string, which might needs be truncated.
:return: splitted String</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.get_gdc_filename_list">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">get_gdc_filename_list</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_to_gdc_manifest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.get_gdc_filename_list" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates a list of GDC files, for which metadata has to be downloaded
:param path_to_gdc_manifest: path to GDC manifest file
:return: returns a unique list of GDC file_names</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.get_icgc_project_list">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">get_icgc_project_list</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_to_dcc_manifest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.get_icgc_project_list" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates a unique list of ICGC DCC projects, for which metadata has to be downloaded
:param path_to_dcc_manifest: Path to ICGC DCC manifest file
:return: list of unique ICGC DCC project codes</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.get_sra_accession_list">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">get_sra_accession_list</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_to_sra_acc_list</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.get_sra_accession_list" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates a unique list of SRA accessions, for which metadata has to be downloaded
:param path_to_sra_acc_list: path to SRA Accession List
:return: list of unique SRA accesssions</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.gunzip_files">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">gunzip_files</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.gunzip_files" title="Permalink to this definition">¶</a></dt>
<dd><p>This method unzips all .gz files in a given directory, and deletes the .gz files afterwards.
:param directory: Path to directory, which contains files to be unzipped
:return: None</p>
</dd></dl>

<dl class="py method">
<dt id="database_preprocessor.Helper.retrieve_data">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">retrieve_data</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">destination</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retry_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.Helper.retrieve_data" title="Permalink to this definition">¶</a></dt>
<dd><p>This method downloads files from a remote source and saves them in the destination directory.
:param url: URL to file, which has to be downloaded
:param destination: Path to directory, where the files should be saved
:param retry_count: When retry_count is 0, stop script
:return:</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="database_preprocessor.ListDict">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">database_preprocessor.</span></code><code class="sig-name descname"><span class="pre">ListDict</span></code><a class="headerlink" href="#database_preprocessor.ListDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt id="database_preprocessor.ListDict.base_path">
<code class="sig-name descname"><span class="pre">base_path</span></code><em class="property"> <span class="pre">=</span> <span class="pre">PosixPath('/home/flo/github/thesis/scripts')</span></em><a class="headerlink" href="#database_preprocessor.ListDict.base_path" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.current_dir">
<code class="sig-name descname"><span class="pre">current_dir</span></code><em class="property"> <span class="pre">=</span> <span class="pre">'/home/flo/github/thesis/docs'</span></em><a class="headerlink" href="#database_preprocessor.ListDict.current_dir" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.database_columns">
<code class="sig-name descname"><span class="pre">database_columns</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{'access':</span> <span class="pre">'consent',</span> <span class="pre">'bio_project':</span> <span class="pre">'project_id',</span> <span class="pre">'bio_sample':</span> <span class="pre">'sample_id',</span> <span class="pre">'file_name':</span> <span class="pre">'run',</span> <span class="pre">'icgc_donor_id':</span> <span class="pre">'donor_id',</span> <span class="pre">'icgc_sample_id':</span> <span class="pre">'sample_id',</span> <span class="pre">'sex':</span> <span class="pre">'donor_sex'}</span></em><a class="headerlink" href="#database_preprocessor.ListDict.database_columns" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.db_col">
<code class="sig-name descname"><span class="pre">db_col</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/column_db_dict.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.db_col" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.don">
<code class="sig-name descname"><span class="pre">don</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/donor_df_col_list.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.don" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.donor_db">
<code class="sig-name descname"><span class="pre">donor_db</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/donor_db_list.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.donor_db" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.donor_db_list">
<code class="sig-name descname"><span class="pre">donor_db_list</span></code><em class="property"> <span class="pre">=</span> <span class="pre">['project_code',</span> <span class="pre">'donor_id',</span> <span class="pre">'submitted_donor_id',</span> <span class="pre">'donor_sex',</span> <span class="pre">'donor_vital_status',</span> <span class="pre">'disease_status_last_followup',</span> <span class="pre">'donor_relapse_type',</span> <span class="pre">'donor_age_at_diagnosis',</span> <span class="pre">'donor_age_at_enrollment',</span> <span class="pre">'donor_age_at_last_followup',</span> <span class="pre">'donor_relapse_interval',</span> <span class="pre">'donor_diagnosis_icd10',</span> <span class="pre">'donor_tumour_staging_system_at_diagnosis',</span> <span class="pre">'donor_tumour_stage_at_diagnosis',</span> <span class="pre">'donor_tumour_stage_at_diagnosis_supplemental',</span> <span class="pre">'donor_survival_time',</span> <span class="pre">'donor_interval_of_last_followup',</span> <span class="pre">'prior_malignancy',</span> <span class="pre">'cancer_type_prior_malignancy',</span> <span class="pre">'cancer_history_first_degree_relative']</span></em><a class="headerlink" href="#database_preprocessor.ListDict.donor_db_list" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.donor_df_cols">
<code class="sig-name descname"><span class="pre">donor_df_cols</span></code><em class="property"> <span class="pre">=</span> <span class="pre">['icgc_donor_id',</span> <span class="pre">'project_code',</span> <span class="pre">'study_donor_involved_in',</span> <span class="pre">'submitted_donor_id',</span> <span class="pre">'donor_sex',</span> <span class="pre">'donor_vital_status',</span> <span class="pre">'disease_status_last_followup',</span> <span class="pre">'donor_relapse_type',</span> <span class="pre">'donor_age_at_diagnosis',</span> <span class="pre">'donor_age_at_enrollment',</span> <span class="pre">'donor_age_at_last_followup',</span> <span class="pre">'donor_relapse_interval',</span> <span class="pre">'donor_diagnosis_icd10',</span> <span class="pre">'donor_tumour_staging_system_at_diagnosis',</span> <span class="pre">'donor_tumour_stage_at_diagnosis',</span> <span class="pre">'donor_tumour_stage_at_diagnosis_supplemental',</span> <span class="pre">'donor_survival_time',</span> <span class="pre">'donor_interval_of_last_followup',</span> <span class="pre">'prior_malignancy',</span> <span class="pre">'cancer_type_prior_malignancy',</span> <span class="pre">'cancer_history_first_degree_relative']</span></em><a class="headerlink" href="#database_preprocessor.ListDict.donor_df_cols" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.gdc_attr">
<code class="sig-name descname"><span class="pre">gdc_attr</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/gdc_attr_list.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.gdc_attr" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.gdc_attributes">
<code class="sig-name descname"><span class="pre">gdc_attributes</span></code><em class="property"> <span class="pre">=</span> <span class="pre">['access',</span> <span class="pre">'experimental_strategy',</span> <span class="pre">'platform',</span> <span class="pre">'type',</span> <span class="pre">'cases.demographic.gender',</span> <span class="pre">'cases.diagnoses.primary_diagnosis',</span> <span class="pre">'cases.diagnoses.tumor_grade',</span> <span class="pre">'cases.diagnoses.tumor_stage',</span> <span class="pre">'cases.diagnoses.vital_status',</span> <span class="pre">'cases.diagnoses.age_at_diagnosis',</span> <span class="pre">'cases.project.dbgap_accession_number',</span> <span class="pre">'cases.project.disease_type',</span> <span class="pre">'cases.project.primary_site',</span> <span class="pre">'cases.project.project_id',</span> <span class="pre">'cases.samples.sample_id',</span> <span class="pre">'cases.samples.sample_type',</span> <span class="pre">'cases.submitter_id',</span> <span class="pre">'cases.diagnoses.last_known_disease_status',</span> <span class="pre">'cases.diagnoses.prior_malignancy',</span> <span class="pre">'cases.diagnoses.classification_of_tumor',</span> <span class="pre">'cases.samples.tumor_code',</span> <span class="pre">'cases.samples.tumor_code_id',</span> <span class="pre">'cases.samples.tumor_descriptor',</span> <span class="pre">'cases.samples.preservation_method',</span> <span class="pre">'cases.samples.sample_type',</span> <span class="pre">'cases.samples.tissue_type',</span> <span class="pre">'cases.samples.tumor_code',</span> <span class="pre">'cases.samples.tumor_code_id']</span></em><a class="headerlink" href="#database_preprocessor.ListDict.gdc_attributes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.gdc_col_dict">
<code class="sig-name descname"><span class="pre">gdc_col_dict</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{'access':</span> <span class="pre">'consent',</span> <span class="pre">'cases.0.demographic.gender':</span> <span class="pre">'donor_sex',</span> <span class="pre">'cases.0.diagnoses.0.age_at_diagnosis':</span> <span class="pre">'donor_age_at_diagnosis',</span> <span class="pre">'cases.0.diagnoses.0.classification_of_tumor':</span> <span class="pre">'classification_of_tumour',</span> <span class="pre">'cases.0.diagnoses.0.last_known_disease_status':</span> <span class="pre">'last_known_disease_status',</span> <span class="pre">'cases.0.diagnoses.0.primary_diagnosis':</span> <span class="pre">'donor_primary_diagnosis',</span> <span class="pre">'cases.0.diagnoses.0.prior_malignancy':</span> <span class="pre">'prior_malignancy',</span> <span class="pre">'cases.0.diagnoses.0.tumor_grade':</span> <span class="pre">'tumour_grade',</span> <span class="pre">'cases.0.diagnoses.0.tumor_stage':</span> <span class="pre">'donor_tumour_stage_at_diagnosis',</span> <span class="pre">'cases.0.project.dbgap_accession_number':</span> <span class="pre">'dbgap_study_accession',</span> <span class="pre">'cases.0.project.disease_type':</span> <span class="pre">'disease_type',</span> <span class="pre">'cases.0.project.primary_site':</span> <span class="pre">'body_site',</span> <span class="pre">'cases.0.project.project_id':</span> <span class="pre">'project_code',</span> <span class="pre">'cases.0.samples.0.preservation_method':</span> <span class="pre">'specimen_processing',</span> <span class="pre">'cases.0.samples.0.sample_id':</span> <span class="pre">'sample_id</span> <span class="pre">',</span> <span class="pre">'cases.0.samples.0.sample_type':</span> <span class="pre">'specimen_type',</span> <span class="pre">'cases.0.samples.0.tissue_type':</span> <span class="pre">'tissue_type',</span> <span class="pre">'cases.0.samples.0.tumor_code':</span> <span class="pre">'tumour_code',</span> <span class="pre">'cases.0.samples.0.tumor_code_id':</span> <span class="pre">'tumour_code_id',</span> <span class="pre">'cases.0.samples.0.tumor_descriptor':</span> <span class="pre">'tumour_descriptor',</span> <span class="pre">'cases.0.submitter_id':</span> <span class="pre">'donor_id',</span> <span class="pre">'experimental_strategy':</span> <span class="pre">'library_strategy',</span> <span class="pre">'id':</span> <span class="pre">'run'}</span></em><a class="headerlink" href="#database_preprocessor.ListDict.gdc_col_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.gdc_cols">
<code class="sig-name descname"><span class="pre">gdc_cols</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/gdc_col_dict.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.gdc_cols" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.icgc_col_names">
<code class="sig-name descname"><span class="pre">icgc_col_names</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/icgc_rep_dict.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.icgc_col_names" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.icgc_col_names_repository">
<code class="sig-name descname"><span class="pre">icgc_col_names_repository</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{'Access':</span> <span class="pre">'access',</span> <span class="pre">'Data</span> <span class="pre">Type':</span> <span class="pre">'data_type',</span> <span class="pre">'Experimental</span> <span class="pre">Strategy':</span> <span class="pre">'experiemntal_strategy',</span> <span class="pre">'File</span> <span class="pre">ID':</span> <span class="pre">'file_id',</span> <span class="pre">'File</span> <span class="pre">Name':</span> <span class="pre">'file_name',</span> <span class="pre">'Format':</span> <span class="pre">'format',</span> <span class="pre">'ICGC</span> <span class="pre">Donor':</span> <span class="pre">'icgc_donor_id',</span> <span class="pre">'Object</span> <span class="pre">ID':</span> <span class="pre">'object_id',</span> <span class="pre">'Project':</span> <span class="pre">'project_code',</span> <span class="pre">'Repository':</span> <span class="pre">'repository',</span> <span class="pre">'Sample</span> <span class="pre">ID':</span> <span class="pre">'icgc_sample_id',</span> <span class="pre">'Size</span> <span class="pre">(bytes)':</span> <span class="pre">'size_bytes',</span> <span class="pre">'Specimen</span> <span class="pre">ID':</span> <span class="pre">'icgc_specimen_id',</span> <span class="pre">'Specimen</span> <span class="pre">Type':</span> <span class="pre">'specimen_type'}</span></em><a class="headerlink" href="#database_preprocessor.ListDict.icgc_col_names_repository" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.icgc_tissue">
<code class="sig-name descname"><span class="pre">icgc_tissue</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/icgc_tissue_dict.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.icgc_tissue" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.icgc_tissue_dict">
<code class="sig-name descname"><span class="pre">icgc_tissue_dict</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{'MALY-DE':</span> <span class="pre">'Blood',</span> <span class="pre">'PACA-CA':</span> <span class="pre">'Pancreas'}</span></em><a class="headerlink" href="#database_preprocessor.ListDict.icgc_tissue_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.rep_man">
<code class="sig-name descname"><span class="pre">rep_man</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/icgc_repository_manifdest_col_list.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.rep_man" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.rep_man_cols">
<code class="sig-name descname"><span class="pre">rep_man_cols</span></code><em class="property"> <span class="pre">=</span> <span class="pre">['access',</span> <span class="pre">'file_id',</span> <span class="pre">'object_id',</span> <span class="pre">'file_name',</span> <span class="pre">'icgc_donor_id',</span> <span class="pre">'icgc_specimen_id',</span> <span class="pre">'specimen_type',</span> <span class="pre">'icgc_sample_id',</span> <span class="pre">'repository',</span> <span class="pre">'project_code',</span> <span class="pre">'Study',</span> <span class="pre">'data_type',</span> <span class="pre">'experiemntal_strategy',</span> <span class="pre">'format',</span> <span class="pre">'size_bytes']</span></em><a class="headerlink" href="#database_preprocessor.ListDict.rep_man_cols" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sam">
<code class="sig-name descname"><span class="pre">sam</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/sample_df_col_list.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sam" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sample_db">
<code class="sig-name descname"><span class="pre">sample_db</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/sample_db_list.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sample_db" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sample_db_list">
<code class="sig-name descname"><span class="pre">sample_db_list</span></code><em class="property"> <span class="pre">=</span> <span class="pre">['run',</span> <span class="pre">'sample_id',</span> <span class="pre">'project_code',</span> <span class="pre">'submitted_sample_id',</span> <span class="pre">'icgc_specimen_id',</span> <span class="pre">'submitted_specimen_id',</span> <span class="pre">'donor_id',</span> <span class="pre">'body_site',</span> <span class="pre">'submitted_donor_id',</span> <span class="pre">'study_specimen_involved_in',</span> <span class="pre">'specimen_type',</span> <span class="pre">'specimen_type_other',</span> <span class="pre">'specimen_interval',</span> <span class="pre">'specimen_donor_treatment_type',</span> <span class="pre">'specimen_donor_treatment_type_other',</span> <span class="pre">'specimen_processing',</span> <span class="pre">'specimen_processing_other',</span> <span class="pre">'specimen_storage',</span> <span class="pre">'specimen_storage_other',</span> <span class="pre">'tumour_confirmed',</span> <span class="pre">'specimen_biobank',</span> <span class="pre">'specimen_biobank_id',</span> <span class="pre">'specimen_available',</span> <span class="pre">'tumour_histological_type',</span> <span class="pre">'tumour_grading_system',</span> <span class="pre">'tumour_grade',</span> <span class="pre">'tumour_grade_supplemental',</span> <span class="pre">'tumour_stage_system',</span> <span class="pre">'tumour_stage',</span> <span class="pre">'tumour_stage_supplemental',</span> <span class="pre">'study_donor_involved_in',</span> <span class="pre">'consent',</span> <span class="pre">'repository',</span> <span class="pre">'experiment',</span> <span class="pre">'sample',</span> <span class="pre">'sample_type',</span> <span class="pre">'sample_name',</span> <span class="pre">'source',</span> <span class="pre">'disease',</span> <span class="pre">'tumour',</span> <span class="pre">'affection_status',</span> <span class="pre">'analyte_type',</span> <span class="pre">'histological_type',</span> <span class="pre">'center_name',</span> <span class="pre">'submission',</span> <span class="pre">'portal']</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sample_db_list" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sample_df_cols">
<code class="sig-name descname"><span class="pre">sample_df_cols</span></code><em class="property"> <span class="pre">=</span> <span class="pre">['icgc_sample_id',</span> <span class="pre">'project_code',</span> <span class="pre">'submitted_sample_id',</span> <span class="pre">'icgc_specimen_id',</span> <span class="pre">'submitted_specimen_id',</span> <span class="pre">'icgc_donor_id',</span> <span class="pre">'submitted_donor_id',</span> <span class="pre">'analyzed_sample_interval',</span> <span class="pre">'percentage_cellularity',</span> <span class="pre">'level_of_cellularity',</span> <span class="pre">'study',</span> <span class="pre">'study_specimen_involved_in',</span> <span class="pre">'specimen_type',</span> <span class="pre">'specimen_type_other',</span> <span class="pre">'specimen_interval',</span> <span class="pre">'specimen_donor_treatment_type',</span> <span class="pre">'specimen_donor_treatment_type_other',</span> <span class="pre">'specimen_processing',</span> <span class="pre">'specimen_processing_other',</span> <span class="pre">'specimen_storage',</span> <span class="pre">'specimen_storage_other',</span> <span class="pre">'tumour_confirmed',</span> <span class="pre">'specimen_biobank',</span> <span class="pre">'specimen_biobank_id',</span> <span class="pre">'specimen_available',</span> <span class="pre">'tumour_histological_type',</span> <span class="pre">'tumour_grading_system',</span> <span class="pre">'tumour_grade',</span> <span class="pre">'tumour_grade_supplemental',</span> <span class="pre">'tumour_stage_system',</span> <span class="pre">'tumour_stage',</span> <span class="pre">'tumour_stage_supplemental',</span> <span class="pre">'digital_image_of_stained_section']</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sample_df_cols" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sra_col">
<code class="sig-name descname"><span class="pre">sra_col</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/sra_col_dict.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sra_col" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sra_col_dict">
<code class="sig-name descname"><span class="pre">sra_col_dict</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{'Affection_Status':</span> <span class="pre">'affection_status',</span> <span class="pre">'Analyte_Type':</span> <span class="pre">'analyte_type',</span> <span class="pre">'AssemblyName':</span> <span class="pre">'assembly_name',</span> <span class="pre">'BioProject':</span> <span class="pre">'project_code',</span> <span class="pre">'BioSample':</span> <span class="pre">'sample_id',</span> <span class="pre">'Body_Site':</span> <span class="pre">'body_site',</span> <span class="pre">'CenterName':</span> <span class="pre">'center_name',</span> <span class="pre">'Consent':</span> <span class="pre">'consent',</span> <span class="pre">'Disease':</span> <span class="pre">'disease',</span> <span class="pre">'Experiment':</span> <span class="pre">'experiment',</span> <span class="pre">'Histological_Type':</span> <span class="pre">'histological_type',</span> <span class="pre">'InsertDev':</span> <span class="pre">'insert_dev',</span> <span class="pre">'InsertSize':</span> <span class="pre">'insert_size',</span> <span class="pre">'LibraryLayout':</span> <span class="pre">'library_layout',</span> <span class="pre">'LibraryName':</span> <span class="pre">'library_name',</span> <span class="pre">'LibrarySelection':</span> <span class="pre">'library_selection',</span> <span class="pre">'LibrarySource':</span> <span class="pre">'library_source',</span> <span class="pre">'LibraryStrategy':</span> <span class="pre">'library_strategy',</span> <span class="pre">'LoadDate':</span> <span class="pre">'load_date',</span> <span class="pre">'Model':</span> <span class="pre">'model',</span> <span class="pre">'Platform':</span> <span class="pre">'plattform',</span> <span class="pre">'ProjectID':</span> <span class="pre">'project_id',</span> <span class="pre">'ReadHash':</span> <span class="pre">'read_hash',</span> <span class="pre">'ReleaseDate':</span> <span class="pre">'release_date',</span> <span class="pre">'Run':</span> <span class="pre">'run',</span> <span class="pre">'RunHash':</span> <span class="pre">'run_hash',</span> <span class="pre">'SRAStudy':</span> <span class="pre">'sra_study',</span> <span class="pre">'Sample':</span> <span class="pre">'sample',</span> <span class="pre">'SampleName':</span> <span class="pre">'sample_name',</span> <span class="pre">'SampleType':</span> <span class="pre">'sample_type',</span> <span class="pre">'Sex':</span> <span class="pre">'sex',</span> <span class="pre">'Study_Pubmed_id':</span> <span class="pre">'study_pubmed_id',</span> <span class="pre">'Subject_ID':</span> <span class="pre">'donor_id',</span> <span class="pre">'Submission':</span> <span class="pre">'submission',</span> <span class="pre">'Tumor':</span> <span class="pre">'tumour',</span> <span class="pre">'avgLength':</span> <span class="pre">'avg_length'}</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sra_col_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sra_df">
<code class="sig-name descname"><span class="pre">sra_df</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;_io.TextIOWrapper</span> <span class="pre">name='/home/flo/github/thesis/scripts/dicts_lists/sra_df_col_list.json'</span> <span class="pre">mode='r'</span> <span class="pre">encoding='UTF-8'&gt;</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sra_df" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.sra_df_cols">
<code class="sig-name descname"><span class="pre">sra_df_cols</span></code><em class="property"> <span class="pre">=</span> <span class="pre">['run',</span> <span class="pre">'release_date',</span> <span class="pre">'load_date',</span> <span class="pre">'spots',</span> <span class="pre">'bases',</span> <span class="pre">'spots_with_mates',</span> <span class="pre">'avg_length',</span> <span class="pre">'size_MB',</span> <span class="pre">'assembly_name',</span> <span class="pre">'download_path',</span> <span class="pre">'experiment',</span> <span class="pre">'library_name',</span> <span class="pre">'library_strategy',</span> <span class="pre">'library_selection',</span> <span class="pre">'library_source',</span> <span class="pre">'library_layout',</span> <span class="pre">'insert_size',</span> <span class="pre">'insert_dev',</span> <span class="pre">'plattform',</span> <span class="pre">'model',</span> <span class="pre">'sra_study',</span> <span class="pre">'project_code',</span> <span class="pre">'study_pubmed_id',</span> <span class="pre">'sample',</span> <span class="pre">'sample_id',</span> <span class="pre">'sample_type',</span> <span class="pre">'TaxID',</span> <span class="pre">'ScientificName',</span> <span class="pre">'sample_name',</span> <span class="pre">'g1k_pop_code',</span> <span class="pre">'source',</span> <span class="pre">'g1k_analysis_group',</span> <span class="pre">'donor_sex',</span> <span class="pre">'disease',</span> <span class="pre">'tumour',</span> <span class="pre">'affection_status',</span> <span class="pre">'analyte_type',</span> <span class="pre">'histological_type',</span> <span class="pre">'body_site',</span> <span class="pre">'center_name',</span> <span class="pre">'submission',</span> <span class="pre">'dbgap_study_accession',</span> <span class="pre">'consent',</span> <span class="pre">'run_hash',</span> <span class="pre">'read_hash']</span></em><a class="headerlink" href="#database_preprocessor.ListDict.sra_df_cols" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.ListDict.stringtie_dict">
<code class="sig-name descname"><span class="pre">stringtie_dict</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{'Coverage':</span> <span class="pre">'coverage',</span> <span class="pre">'End':</span> <span class="pre">'end_pos',</span> <span class="pre">'FPKM':</span> <span class="pre">'fpkm',</span> <span class="pre">'Gene</span> <span class="pre">ID':</span> <span class="pre">'gene_id',</span> <span class="pre">'Gene</span> <span class="pre">Name':</span> <span class="pre">'gene_name',</span> <span class="pre">'Reference':</span> <span class="pre">'reference',</span> <span class="pre">'Start':</span> <span class="pre">'start_pos',</span> <span class="pre">'Strand':</span> <span class="pre">'strand',</span> <span class="pre">'TPM':</span> <span class="pre">'tpm'}</span></em><a class="headerlink" href="#database_preprocessor.ListDict.stringtie_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="database_preprocessor.Log">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">database_preprocessor.</span></code><code class="sig-name descname"><span class="pre">Log</span></code><a class="headerlink" href="#database_preprocessor.Log" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt id="database_preprocessor.Log.ch">
<code class="sig-name descname"><span class="pre">ch</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;StreamHandler</span> <span class="pre">&lt;stderr&gt;</span> <span class="pre">(NOTSET)&gt;</span></em><a class="headerlink" href="#database_preprocessor.Log.ch" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.Log.formatter">
<code class="sig-name descname"><span class="pre">formatter</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;logging.Formatter</span> <span class="pre">object&gt;</span></em><a class="headerlink" href="#database_preprocessor.Log.formatter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="database_preprocessor.Log.logger">
<code class="sig-name descname"><span class="pre">logger</span></code><em class="property"> <span class="pre">=</span> <span class="pre">&lt;Logger</span> <span class="pre">Expression</span> <span class="pre">Store</span> <span class="pre">Script</span> <span class="pre">Logger</span> <span class="pre">(INFO)&gt;</span></em><a class="headerlink" href="#database_preprocessor.Log.logger" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="database_preprocessor.SQLscripts">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">database_preprocessor.</span></code><code class="sig-name descname"><span class="pre">SQLscripts</span></code><a class="headerlink" href="#database_preprocessor.SQLscripts" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="database_preprocessor.SQLscripts.prepare_populate_script">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">prepare_populate_script</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_dest</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.SQLscripts.prepare_populate_script" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates a populate.sql script based on the given input of db_dest
:param db_dest: Path do directory where files for database are stored
:return: None</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="database_preprocessor.main">
<code class="sig-prename descclassname"><span class="pre">database_preprocessor.</span></code><code class="sig-name descname"><span class="pre">main</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#database_preprocessor.main" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the main class, which performs all necessary steps to successfully download, process and export metadata
and data for the database</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rnaseq</strong> – Path to the base directory of nf-core/rnaseq</p></li>
<li><p><strong>sra_file</strong> – Path to the SRA accession list txt-file</p></li>
<li><p><strong>dcc_manifest</strong> – Path to the manifest file of ICGC DCC</p></li>
<li><p><strong>gdc_manifest</strong> – Path to the manifest file of GDC</p></li>
<li><p><strong>metadata_dest</strong> – Path to a previously created directory, to store downloaded metadata</p></li>
<li><p><strong>db_dest</strong> – Path to the destination directory of the created csv-files, which are imported to the database</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="modules.html" class="btn btn-neutral float-left" title="scripts" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Florian Kraus.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>